python/ray/data/datasource/iceberg_datasource.py

```diff
@@ -0,0 +1,187 @@
+"""
+Module to read an iceberg table into a Ray Dataset, by using the Ray Datasource API.
+"""
+
+import logging
+from typing import TYPE_CHECKING, Any, Iterable, List, Optional, Tuple, Union
+
+from ray.data import ReadTask, block
+from ray.data.datasource import Datasource
+from ray.util.annotations import DeveloperAPI
+
+if TYPE_CHECKING:
+    from pyiceberg.catalog import Catalog
+    from pyiceberg.expressions import BooleanExpression
+    from pyiceberg.table import DataScan, FileScanTask, Schema, Table
+
+logger = logging.getLogger(__name__)
+
+
+@DeveloperAPI
+class IcebergDatasource(Datasource):
+    """
+    Iceberg datasource to read Iceberg tables into a Ray Dataset. This module heavily
+    uses PyIceberg to read iceberg tables. All the routines in this class override
+    `ray.data.Datasource`.
+    """
+
+    def __init__(
+        self,
+        table_identifier: str,
+        catalog_type: str = "glue",
+        row_filter: Union[str, "BooleanExpression"] = None,
+        selected_fields: Tuple[str, ...] = ("*",),
+        snapshot_id: Optional[int] = None,
+        scan_kwargs: Optional[dict[str, str]] = None,
+        catalog_kwargs: Optional[dict[str, Any]] = None,
+    ):
+        """
+        Initialize an IcebergDatasource.
+
+        Args:
+            table_identifier: Fully qualified table identifier (i.e.,
+                "db_name.table_name")
+            catalog_type: The type of catalog to use PyIceberg with (defaults to
+                "glue")
+            row_filter: A PyIceberg BooleanExpression to use to filter the data *prior*
+                 to reading
+            selected_fields: Which columns from the data to read, passed directly to
+                PyIceberg's load functions
+            snapshot_id: Optional snapshot ID for the Iceberg table
+            scan_kwargs: Optional arguments to pass to PyIceberg's Table.scan()
+                function
+            catalog_kwargs: Optional arguments to use when setting up the Iceberg
+                catalog
+        """
+        from pyiceberg.expressions import AlwaysTrue
+
+        self._scan_kwargs = scan_kwargs if scan_kwargs is not None else {}
+
+        self.table_identifier = table_identifier
+
+        self._catalog_type = catalog_type
+        self._catalog_kwargs = catalog_kwargs if catalog_kwargs is not None else {}
+
+        self._row_filter = row_filter if row_filter is not None else AlwaysTrue()
+        self._selected_fields = selected_fields
+        self._snapshot_id = snapshot_id
+
+    def _get_catalog(self) -> "Catalog":
+        from pyiceberg import catalog
+
+        if self._catalog_type == "glue":
+            # Glue catalogs have no name, and no conf options are needed as long as AWS
+            # credentials are set up
+            return catalog.load_glue(name="", conf=self._catalog_kwargs)
```

Comment:
I would highly recommend using the `load_catalog` method instead since it offers much higher flexibility.

```suggestion
            if "type" not in self._catalog_kwargs:
                self._catalog_kwargs["type"] = self._catalog_type
            return catalog.load_catalog(name="default", **self._catalog_kwargs)
```

Also, setting a name allows the user to pass in configuration [through environment variables, or a `.pyiceberg.yaml` config](https://py.iceberg.apache.org/configuration/).